lib/fdtable.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief File descriptor table
+ *
+ * This file provides generic file descriptor table implementation, suitable
+ * for any I/O object implementing POSIX I/O semantics (i.e. read/write +
+ * aux operations).
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <misc/fdtable.h>
+
+struct fd_entry {
+	void *obj;
+	const struct fd_op_vtable *vtable;
+};
+
+#ifdef CONFIG_POSIX_API
+static struct fd_op_vtable stdinout_fd_op_vtable;
+#endif
+
+static struct fd_entry fdtable[CONFIG_POSIX_MAX_FDS] =
+{
+#ifdef CONFIG_POSIX_API
+	/*
+	 * Predefine entries for stdin/stdout/stderr. Object pointer
```

Comment:
can we perhaps define a constant for unused object pointer values?

Reply:
> can we perhaps define a constant for unused object pointer values?

Well, I don't know how useful that would, I'd think it's rather a once-off exception. Values here were just selected to be distinct and more all less distinguishable if they appear e.g. in crash dump or in gdb output.

Reply:
Defines would be more proper, we should avoid using magic constants as they are just confusing when trying to read code.

Reply:
Update: as I'm forced to define a constant for "reserved" entry, mass-converting this case too.